evmrpc/simulate.go

```diff
@@ -295,16 +295,11 @@ func (b *Backend) StateAtTransaction(ctx context.Context, block *ethtypes.Block,
 		// set block context time as of the block time (block time is the time of the CURRENT block)
 		blockContext.Time = block.Time()
 
-		// set address association for the sender if not present. Note that here we take the shortcut
-		// of querying from the latest height with the assumption that if this tx has been processed
-		// at all then its association must be present in the latest height
-		_, associated := b.keeper.GetSeiAddress(statedb.Ctx(), msg.From)
-		if !associated {
-			seiAddr, associatedNow := b.keeper.GetSeiAddress(b.ctxProvider(LatestCtxHeight), msg.From)
-			if !associatedNow {
-				return nil, vm.BlockContext{}, nil, nil, fmt.Errorf("address %s is not associated in the latest height", msg.From.Hex())
-			}
-			if err := ante.NewEVMPreprocessDecorator(b.keeper, b.keeper.AccountKeeper()).AssociateAddresses(statedb.Ctx(), seiAddr, msg.From, nil); err != nil {
+		// check if this transaction caused legacy address association and simulate if so
+		addressThen := b.keeper.GetSeiAddress(statedb.Ctx(), msg.From)
+		addressNow := b.keeper.GetSeiAddress(b.ctxProvider(LatestCtxHeight), msg.From)
+		if !addressThen.Equals(addressNow) {
+			if err := ante.NewEVMPreprocessDecorator(b.keeper, b.keeper.AccountKeeper()).AssociateAddresses(statedb.Ctx(), addressNow, msg.From, nil); err != nil {
```

Comment:
this is for legacy association created before this change. `addressThen` is the sei address before THAT legacy association happened. What we are doing here is basically checking if this tx, back in time, caused a legacy association, and reproduce that if so.

I think the var name might be what's confusing: `addressNow` doesn't mean the current way of address association, rather it means the association of the address in the latest DB, and it could be a legacy one

Reply:
ahh gotcha, yeah the `addressNow` name was confusing me